README.md
# vcard-creator

[![npm version](https://badge.fury.io/js/vcard-creator.svg)](https://badge.fury.io/js/vcard-creator)
[![jest](https://jestjs.io/img/jest-badge.svg)](https://github.com/facebook/jest)
![tests](https://github.com/joaocarmo/vcard-creator/workflows/Tests/badge.svg)
[![Total alerts](https://img.shields.io/lgtm/alerts/g/joaocarmo/vcard-creator.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/joaocarmo/vcard-creator/alerts/)
[![Language grade: JavaScript](https://img.shields.io/lgtm/grade/javascript/g/joaocarmo/vcard-creator.svg?logo=lgtm&logoWidth=18)](https://lgtm.com/projects/g/joaocarmo/vcard-creator/context:javascript)

A JavaScript vCard 3.0 creator library for both node.js and the web.
It outputs the vCard text that should be saved as a `*.vcf` file.

## Origin

This is based on _jeroendesloovere_'s [vCard][jeroendesloovere] for PHP.

## Installation

```sh
yarn add vcard-creator

# or

npm install vcard-creator
```

## Usage

### As an ESM module (web)

Load **vcard-creator** directly from [skypack][skypack] (CDN).

```html
<script type="module">
  import VCard from 'https://cdn.skypack.dev/vcard-creator'
</script>
```

Demo available in [codepen][codepen].

### On the web (self-hosted)

It's exposed through the _window_ global object as explained below.

`index.html`

```html
<head>
  <script type="text/javascript" src="./js/vcard-creator.js"></script>
  <script type="text/javascript" src="./js/main.js"></script>
</head>
```

`main.js`

```js
// Define a new vCard
var VCard = window.vcardcreator.default
var myVCard = new VCard()

// ...rest of the code
```

### With a bundler / Node.js

With a bundler (e.g. webpack) or in Node.js you can just require / import it.

```js
const VCard = require('vcard-creator').default

// Define a new vCard
const myVCard = new VCard()
```

Or...

```js
import VCard from 'vcard-creator'

// Define a new vCard
const myVCard = new VCard()
```

### Including an image

You need to provide the image already properly encoded (base64). Note that most
applications will probably ignore a photo URL, even if it adheres to the
specification.

```js
// Example in Node.js

const fs = require('fs')
const VCard = require('vcard-creator').default

const imagePath = './path/to/my/assets/sample.jpg'
const image = fs.readFileSync(imagePath, { encoding: 'base64', flag: 'r' })

const vCard = new VCard()

vCard.addPhoto(image, 'JPEG')
```

Include the proper [MIME type][mime-types] (defaults to `JPEG`).

### iCalendar format

For Apple devices that don't support the `vcf` file format, there is a
workaround. Specify the format of the output as `vcalendar` and then save it
with a `ics` file extension instead.

The trick is to create an iCalendar file with a vCard attached.

```js
// Define a new vCard as 'vcalendar'
const myVCalendar = new VCard('vcalendar')

// ...or set it afterwards
const myOtherVCalendar = new VCard()
myOtherVCalendar.setFormat('vcalendar')
```

## Example

```js
import VCard from 'vcard-creator'

// Define a new vCard
const myVCard = new VCard()

// Some variables
const lastname = 'Desloovere'
const firstname = 'Jeroen'
const additional = ''
const prefix = ''
const suffix = ''

myVCard
  // Add personal data
  .addName(lastname, firstname, additional, prefix, suffix)
  // Add work data
  .addCompany('Siesqo')
  .addJobtitle('Web Developer')
  .addRole('Data Protection Officer')
  .addEmail('info@jeroendesloovere.be')
  .addPhoneNumber(1234121212, 'PREF;WORK')
  .addPhoneNumber(123456789, 'WORK')
  .addAddress(null, null, 'street', 'worktown', null, 'workpostcode', 'Belgium')
  .addSocial('https://twitter.com/desloovere_j', 'Twitter', 'desloovere_j')
  .addURL('http://www.jeroendesloovere.be')

console.log(myVCard.toString())
```

Output

```txt
BEGIN:VCARD
VERSION:3.0
REV:2017-08-31T17:00:15.850Z
N;CHARSET=utf-8:Desloovere;Jeroen;;;
FN;CHARSET=utf-8:Jeroen Desloovere
ORG;CHARSET=utf-8:Siesqo
TITLE;CHARSET=utf-8:Web Developer
ROLE;CHARSET=utf-8:Data Protection Officer
EMAIL;INTERNET:info@jeroendesloovere.be
TEL;PREF;WORK:1234121212
TEL;WORK:123456789
ADR;WORK;POSTAL;CHARSET=utf-8:name;extended;street;worktown;state;workpos
 tcode;Belgium
X-SOCIALPROFILE;type=Twitter;x-user=desloovere_j:https://twitter.com/desl
oovere_j
URL:http://www.jeroendesloovere.be
END:VCARD
```

## Forking / Contributing

If you're interested in the development of this project, you can run some ready
to use commands to compile and test your changes.

```sh
# Build
yarn build

# Test
yarn test:unit

yarn test:functional

yarn test:web-build

yarn test:web-export
```

<!-- References -->

[codepen]: https://codepen.io/joaocarmo/pen/PozdprL
[jeroendesloovere]: https://github.com/jeroendesloovere/vcard
[mime-types]: https://www.iana.org/assignments/media-types/media-types.xhtml#image
[skypack]: https://skypack.dev


CHANGELOG.md
# Changelog

## 0.6.0

> Jan 1, 2023

- Adds support for `addNickname`
  ([#38](https://github.com/joaocarmo/vcard-creator/pull/38)) and `addSocial`
  ([#39](https://github.com/joaocarmo/vcard-creator/pull/39))

## 0.5.0

> Aug 11, 2022

- Enhances the types definitions
- Reorganizes the files
- Adds support for `UID` ([#32](https://github.com/joaocarmo/vcard-creator/issues/32))

## 0.4.2

> Jan 9, 2022

- Fixes the issue with the `addPhoneNumber()` method only accepting `number` as
  a parameter, making it impossible to pass leading zeros ([#27](https://github.com/joaocarmo/vcard-creator/issues/27))

## 0.4.1 (no release)

> Jul 10, 2021

- Updated the docs

## 0.4.1

> Mar 25, 2021

- Fixes the `addMediaContent()` method ([#17](https://github.com/joaocarmo/vcard-creator/pull/17) thanks to @Falklian)
- Fixes the `addMediaURL()` method

## 0.4.0

> Mar 21, 2021

- Removed `@deprecated` from the `toString()` method
- Added a custom error object `VCardException`
- Removed the `include` argument from the `addLogo()` and `addPhoto()` methods
- Removed unnecessary comments
- Added the missing functionality to the `addLogo()` and `addPhoto()` methods
- Added the missing functionality to the `addLogoURL()` and `addPhotoURL()` methods

## 0.3.4

> Mar 21, 2021

- Fixes the exported typings ([#13](https://github.com/joaocarmo/vcard-creator/issues/13)) to have the same name as the distribution file

## 0.3.3

> Jan 31, 2021

- Fixes the exported typings ([#13](https://github.com/joaocarmo/vcard-creator/issues/13))

## 0.3.2

> Jan 9, 2021

- Fixes the [subpath exports](https://nodejs.org/api/packages.html#packages_subpath_exports)

## 0.3.1

> Jan 9, 2021

- Updated the TypeScript declaration file's output to match the name of the
  bundle (#11)

## 0.3.0 (no release)

> Nov 6, 2020

- Updated the documentation to include information about using the module in the
  browser (ESM)
- Updated the `package.json` to comply with ESM best practices
- Updated the dependencies

## 0.3.0

> Aug 9, 2020

- Added support for iCalendar data files

## 0.2.0

> Aug 8, 2020

- Converted to TypeScript
- Added unit and functional tests
- Updated the documentation (#1)

## 0.1.0 (no release)

> Aug 11, 2019

- Updated the dependencies to resolve the security vulnerability
  [CVE-2019-10744](https://github.com/lodash/lodash/pull/4336)

## 0.1.0

> Apr 1, 2019

- Changed the named export to a default export, making it incompatible with
  previous versions [BREAKING]
- Renamed the global object from `vcard_creator` to `vcardcreator` [BREAKING]
- Renamed the constructor `VCard` [BREAKING]

## 0.0.6

> Feb 8, 2019

- Updated the dependencies to resolve the security vulnerability in
  [lodash](https://nvd.nist.gov/vuln/detail/CVE-2018-16487)

## 0.0.5

> Jun 1, 2018

- Added tests for CLI, compiled for web (i.g. webpack) and not compiled (i.g.
  used from the global _window_ object)
- Updated the [README.md](README.md)

## 0.0.4

> Apr 26, 2018

- Exposed the library as a compiled bundle to avoid dependencies

## 0.0.3

> Apr 25, 2018

- Removed the _index.js_ and exposed the module directly from the _lib_ folder

## 0.0.2

> Sep 1, 2017

- Removed the _transliteration_ dependency in order to be compatible with
  _UglifyJS_

## 0.0.1

> Aug 31, 2017

- Removed the _locutus_ dependency
- Fixed the folding of long lines into multiple lines

## 0.0.0

> Aug 29, 2017

- Initial release


test-functional/test-vcard.js
#!/usr/bin/env node
var fs = require('fs')
var VCard = require('../dist/vcard-creator').default

// sample image
var imagePath = './lib/assets/sample.jpg'

// read image (base64 enconded)
var image = fs.readFileSync(imagePath, { encoding: 'base64', flag: 'r' })

// define vCard
var vCard = new VCard()

// define variables
var lastname = 'Desloovere'
var firstname = 'Jeroen'
var additional = ''
var prefix = ''
var suffix = ''

vCard
  // add personal data
  .addName(lastname, firstname, additional, prefix, suffix)
  .addNickname('Jero')
  // add work data
  .addCompany('Siesqo')
  .addJobtitle('Web Developer')
  .addRole('Data Protection Officer')
  .addEmail('info@jeroendesloovere.be')
  .addPhoneNumber(1234121212, 'PREF;WORK')
  .addPhoneNumber(123456789, 'WORK')
  .addAddress(
    'name',
    'extended',
    'street',
    'worktown',
    'state',
    'workpostcode',
    'Belgium',
  )
  .addSocial('https://twitter.com/twitteruser', 'Twitter', 'twitteruser')
  .addSocial('http://www.facebook.com/facebookuser', 'Facebook', 'facebookuser')
  .addSocial('http://www.flickr.com/photos/flickruser', 'Flickr', 'flickruser')
  .addSocial(
    'http://www.linkedin.com/in/linkedinuser',
    'LinkedIn',
    'linkedinuser',
  )
  .addSocial('http://www.custom.social/customuser', 'Custom', 'customuser')
  .addURL('http://www.jeroendesloovere.be')
  .addPhoto(image, 'JPEG')

console.log(vCard.toString())
